_posts/2022-9-5-Diplomat @ Google.md

```diff
@@ -0,0 +1,503 @@
+and renamed lifetimes:
+```rust
+#[diplomat::bridge]
+mod ffi {
+    #[diplomat::opaque]
+    pub struct Opaque {
+        // ...
+    }
+
+    pub struct Input<'i> {
+        data: &'i Opaque,
+    }
+
+    pub struct Output<'o> {
+        data: &'o Opaque,
+    }
+
+    impl<'a> Input<'a> {
+        fn get_data<'b>(input: Input<'a>) -> Output<'b>
+        where
+            'a: 'b,
+        {
+            let out = Output { data: input.data }
+            out
+        }
+    }
+}
+```
+The first thing that Diplomat does when it encounters `Input::get_data` is
+examine the return type signature, `Output<'b>`. Since it's a non-opaque struct,
+we look in the declaration of `Output` and store a mapping
+saying that the position of `'o` in the context of `Output`'s declaration
+corresponds with the position of `'b` in `Output<'b>` in the method. Next, we
+see the `data: &'o Opaque` field, and use the mapping to determine that it's
+actually `&'b Opaque` in the context of the method. We then remark that
+`out.data` relies on `'b`, see that there's no more fields to explore in the
+field traversal, and finish.
+
+Once we know which lifetimes to look out for (just `'b`), we traverse the types
+of the
+inputs, which consists of just `Input<'a>`. Since `Input` is a non-opaque struct,
+we look in the declaration of `Input` and store a mapping saying that the
+position of `'i` in the context of `Input`'s declaration corresponds with the
+position of `'a` in `Input<'a>` in the method. Inside of `Input`s declaration
+we see the `data: &'i Opaque` field, so we map the lifetime `'i` back to `'a`,
+and then perform a depth-first search in the lifetime graph to find all shorter
+lifetimes.  This search results in finding `'b`. After confirming that `'b`
+appears in the output, we remark that `input.data` is one of the input fields
+that lives at least `'b`.
+
+In code generation, we put these pieces of information together to determine
+that `out.data` must be outlived by `input.data`, which is exactly what we
+were aiming for.
+
+# Results
+That's a lot of details, and unfortunately it's not 100% integrated yet.
+Thus far, the lifetime coercion detection using depth-first search
+is entire implemented, but tracking borrows on a field-by-field basis is only
+implemented in the JavaScript backend. Aside from not understanding the full
+scope of the problem before starting, the biggest challenge was that Diplomat's
+abstract representation was not designed with lifetimes in mind, and also
+wasn't condusive to writing robust code.
+
+For example, traversing down structs recursively and maintaining a mapping from
+lifetimes in a declaration to lifetimes in the method was bolted on at the end
+after I'd already written a significant amount of code. Additionally, Diplomat's
+model allows for a bunch of illegal states like an opaque struct passed by value.
+Obviously there are validity checks, the loose type system eventaully pollutes
```

Comment:
I think it would be valuable to mention that you designed and implemented the HIR, even if you didn't integrate it fully. I know you talk about it more later. I know you tlak about it more later, but perhaps mention something like " I started work on something to fix this that I'll cover in more detail below"